hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,216 @@
+| `spi_active(spit_t *obj)` | removed | as the async call back is now always invoked on async operation termination (unless cancelled), this status can be tracked from driver layer without any hal request.
+| `uint32_t spi_irq_handler_asynch(spi_t *obj)` | removed | as the event is now passed as an argument to the callback this method is no longer required.
+
+### The new API
+
+```c
+typedef struct {
+    /** Minimum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    minimum_frequency;
+    /** Maximum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    maximum_frequency;
+    /** Each bit represents the corresponding word length. lsb => 1bit, msb => 32bit. */
+    uint32_t    word_length;
+    bool        support_slave_mode; /**< If true, the device can handle SPI slave mode. */
+} spi_capabilities_t;
+
+typedef enum _spi_mode_t {
+  SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_LOW_SAMPLE_SECOND_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_SECOND_EDGE,
+} spi_mode_t;
+
+typedef enum _spi_bit_ordering_t {
+  SPI_BIT_ORDERING_MSB_FIRST,
+  SPI_BIT_ORDERING_LSB_FIRST,
+} spi_bit_ordering_t;
+
+typedef void (*spi_async_handler_f)(spi_t *obj, void *ctx, spi_async_event_t event);
+
+SPIName spi_get_module(PINName MISO, PINName MOSI, PINName MCLK);
+void spi_get_capabilities(SPIName name, PINName SS, spi_capabilities_t *cap);
+
+void spi_init(spi_t *obj, bool is_slave, PINName MISO, PINName MOSI, PINName MCLK, PINName SS);
+void spi_format(spi_t *obj, uint8_t bits, spi_mode_t mode, spi_bit_ordering_t bit_ordering);
+uint32_t spi_frequency(spi_t *obj, uint32_t hz);
+void spi_transfer(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol);
+bool spi_transfer_async(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol, spi_async_handler_f handler, void *ctx, DMAUsage hint);
+void spi_transfer_async_abort(spi_t *obj);
+void spi_free(spi_t *obj);
+```
+
+## Behaviours
+### Defined Behaviours
+
+- `spi_get_module()` returns the `SPIName` unique identifier to the peripheral associated to this SPI channel.
+- `spi_get_capabilities()` fills the given `spi_capabilities_t` instance
+- `spi_get_capabilities()` should consider the `SS` pin when evaluation the `support_slave_mode` capability.  
+  If the given `SS` pin cannot be managed by hardware in slave mode, `support_slave_mode` should be false.
+- At least a symbol width of 8bit must be supported.
+- The supported frequency range must include the range [0.2..2] MHz.
+- The shortest part of the duty cycle must not be shorter than 50% of the expected period.
+- `spi_init()` initializes the pins. 
```

Comment:
> Considering the many changes that are proposed here, why not remove this ?

What do you refer to by "this" ?

>  is there a mandatory list of calls to init , format, set_frequency ... before the HW isinitialized.

I updated the document to add details about that. With the new details, at least one call of each method is required before HW is initialized. However, this is open to discussion as it would also make sense to have spi_init to set up default parameters and then eventually calling `spi_format` and `spi_frequency` to change specific parameters.